### ISSUE TITLE
learning rate vs eta - master vs 0.47

### ISSUE THREAD

**Issue by gugatr0n1c** [dmlc/xgboost#1054]
Hello,

I am not sure, what is correct behavior but:

Python version (win and linux):

in 0.47 release (checkpoint before xgboost-brick) and older:
- setting learning_rates as list in train() is not working (I guess i should work at least at 0.47 right - according to release notes?)
- only working is bst:eta as parametr (float scalar)
  --> this I guess correct behavior... 

in master branch:
- learning_rate in train() is working correctly (tested for list)
- but setting bst:eta = 0.002 is not working (learning_rates in train() is not set), seems to me it takes only default value

So now I need to specify constant learning_rate only as list of constant values in train()?

Actually it is no so big issue, but older scripts specifing bst:eta as tree parametrs are not compatible with master branch...

Or I am missing something? Because docs are telling still bst:eta = something is only correct way...

thx

**Comment by tqchen** [comment]
The bst prefix is now deprecated, remove bst prefix shoulw ork

**Comment by tqchen** [comment]
Thanks for point out the outdate of the doc, can you submit a PR to fix this problem?

**Comment by gugatr0n1c** [comment]
@tqchen
thx for pointing out on prefix 'bst:' - I missed that.. now it is working

I did some more experiments before PR and I have some questions:

1]
I am running some code with 
a] eta: 0.01 as tree parameters and not set learning_rate in train()
b] not set eta, but set learning_rates in train() to list: lr_list = [0.01] * iteration_count

And I got slicy different result (but signifacant over stochasticity of xgboost).. Better result are for case a] (running several times)
Is it implementation of learning_rates same? I have no idea why it is different... Only tip is first iteration or different for gblinear

2] in docs there are tree specific parameters, but actually they are working for gblinear as well:
a] eta
b] subsample
I assume this parameteres should be noticed for gblinear as well..?

**Comment by Far0n** [comment]
1] is not supposed to happen, but I see the same effect.

As soon as you call _LIB.XGBoosterSetParam before a boosting round, the results begin to differ. For example: If you call bst.set_param('foobar',0) before bst.update(dtrain, i, obj) in python, the result is different. @tqchen any idea why that is happening?

**Comment by khotilov** [comment]
Most likely, it's because random seed is reset every time Configure is called:
https://github.com/dmlc/xgboost/blob/b3c9e6a0db0a7eb755949ac6b26e3ef805738350/src/learner.cc#L194 
With the same seed every iteration, subsampling would be picking the same sets - and performance would suffer.

An easy check might be to turn the `seed_per_iteration` on
https://github.com/dmlc/xgboost/blob/b3c9e6a0db0a7eb755949ac6b26e3ef805738350/src/learner.cc#L285

But I cannot think of any well motivated need for the seed resetting within Configure after a model is initiated. I can submit a PR.

**Comment by khotilov** [comment]
Yep, I can see the same random numbers at every iteration when running in python.

I didn't see it in R as it uses a custom random generation engine seeded from R, so the seed method is dummy in there.

**Comment by Far0n** [comment]
thx for pointing that out @khotilov. I would like to see a PR for that issue. If that change is not wanted at the C++ core level, I guess the only workaround would be to set a different seed in thy py-wrapper for each iteration.

**Comment by tqchen** [comment]
I think not resetting seed everytime reconfigure happens is reasonable @khotilov a PR is welcomed
